### PR TITLE
INFRA-1438: support native heka packs in firehose output

### PR DESCRIPTION
**Pull request by burnsed**
This change removes the now unused json_encoder. It also adds support for the firehose output to handle native heka packs as part of the output itself.

Tested using `heka-testing`  and writing to `staging-il-clever-com-router-haproxy`

### CODE REVIEW

File: firehose_output.go [hunk 1]
```diff
@@ -32,16 +33,29 @@ func (f *FirehoseOutput) Init(rawConfig interface{}) error {
...
+	object["payload"] = m.GetPayload()
+	object["envversion"] = m.GetEnvVersion()
+	object["pid"] = m.GetPid()
+	object["hostname"] = m.GetHostname()
+
+	// store each dynamic field as a top level entry
+	for _, field := range m.Fields {
+		// ignore byte fields
+		if field.GetValueType() != message.Field_BYTES {
+			object[*field.Name] = field.GetValue()
```
Comment (nathanleiby) [inline firehose_output.go]: should check if `*field.Name` is nil, I think

**Comment by dfuentes** [comment]
lgtm

File: firehose_output.go [hunk 2]
```diff
@@ -32,36 +33,48 @@ func (f *FirehoseOutput) Init(rawConfig interface{}) error {
...
 		payload := pack.Message.GetPayload()
 		timestamp := time.Unix(0, pack.Message.GetTimestamp()).Format("2006-01-02 15:04:05.000")
-		pack.Recycle(nil)
 
 		// Verify input is valid json
 		object := make(map[string]interface{})
 		err := json.Unmarshal([]byte(payload), &object)
 		if err != nil {
-			// attempt to use Encoder if it exists
-			if encode_err := f.useEncoder(or, pack, &object); encode_err != nil {
```
Comment (mohit) [inline firehose_output.go]: what was the issue with using the encoder?

File: firehose_output.go [hunk 3]
```diff
@@ -32,36 +33,48 @@ func (f *FirehoseOutput) Init(rawConfig interface{}) error {
...
-
-	return json.Unmarshal(outBytes, &object)
+	return object
 }
 
 func (f *FirehoseOutput) Run(or pipeline.OutputRunner, h pipeline.PluginHelper) error {
 	for pack := range or.InChan() {
 		payload := pack.Message.GetPayload()
 		timestamp := time.Unix(0, pack.Message.GetTimestamp()).Format("2006-01-02 15:04:05.000")
-		pack.Recycle(nil)
```
Comment (mohit) [inline firehose_output.go]: I see this was causing issues :/

**Pull request merged**